Add throttle_average sensor filter

sermayoral opened the pull request:
# What does this implement/fix? 

A way to get the average of a sensor periodically. For example, a thermometer that reports data every 120 seconds, and that the reported data is the average of all the values read in that period.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/569

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1521

## Test Environment

- [X] ESP32
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
sensor:
  - platform: xiaomi_lywsdcgq
    mac_address: 58:2D:34:38:3D:CB
    temperature:
      name: "Kitchen Temperature"
      accuracy_decimals: 2
      filters:
        - throttle_average: 120s
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

sermayoral committed:
First version of periodical_average filter sensor

sermayoral committed:
Initialize values in each iteration

sermayoral committed:
Fix lint

sermayoral committed:
Fix lint

sermayoral committed:
Fix lint

oxan commented on esphome/components/sensor/filter.cpp:
```diff
@@ -310,6 +310,29 @@ void HeartbeatFilter::setup() {
+  });
+}
+float PeriodicalAverageFilter::get_setup_priority() const { return setup_priority::HARDWARE; }
+
```
I think it makes more sense to put this filter with the sliding/exponential moving average filters.

sermayoral replied:
Do you mean to move the code just below the filters you have mentioned?

oxan replied:
Yes.

oxan commented on esphome/components/sensor/filter.h:
```diff
@@ -272,6 +272,22 @@ class HeartbeatFilter : public Filter, public Component {
+ protected:
+  uint32_t time_period_;
+  float sum_{0.0f};
+  size_t n_{0};
```
This can be a regular (unsigned) int, `size_t` is usually only used for the size of objects, buffers and arrays.

sermayoral replied:
You are right

sermayoral committed:
Replace size_t with unsigned int to accumulated values in periodical average filter sensor

sermayoral committed:
Put the periodical_average filter with the sliding/exponential moving average filters

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -187,6 +187,29 @@ optional<float> ExponentialMovingAverageFilter::new_value(float value) {
+
+optional<float> PeriodicalAverageFilter::new_value(float value) {
+  ESP_LOGVV(TAG, "PeriodicalAverageFilter(%p)::new_value(value=%f)", this, value);
+  this->sum_ += value;
```
This should check if `value` is nan, and if so not add it (otherwise all the sum gets messed up)

sermayoral replied:
You're right :-)

OttoWinter commented on esphome/components/sensor/filter.cpp:
```diff
@@ -187,6 +187,29 @@ optional<float> ExponentialMovingAverageFilter::new_value(float value) {
+void PeriodicalAverageFilter::setup() {
+  this->set_interval("periodical_average", this->time_period_, [this]() {
+    ESP_LOGVV(TAG, "PeriodicalAverageFilter(%p)::interval(sum=%f, n=%i)", this, this->sum_, this->n_);
+    if (this->n_ == 0)
```
If no sample occured in that time interval I think we should not silently drop that.

IMO, Let's send NAN in that case (if the user does want that dropped, then they can use the filter out nan filter)

sermayoral replied:
Seem right. @rmeissn we should specify it in the documentation.

sermayoral replied:
@OttoWinter the change is done. But I asked myself, What happen in Home Assistant when the filter sends NaN value? Does the HA sensor become unavailable?

sermayoral committed:
Ensure a new value is not NaN. To send NaN when no sample ocurred in the time interval

sermayoral committed:
Fix non-void lambda does not return a value in all control paths

sermayoral committed:
Rename periodical_average to throttle_average

sermayoral committed:
Fix test1.yaml renaming periodical_average to throttle_average

sermayoral commented:
I have renamed the component from `periodical_average` to `throttle_average` according to #569 

if the reviewers think the old name is better, I just go back to the previous commit. If there are other proposals, it is studied and changed if necessary.

Thanks!
